src/helpers.rs

```diff
@@ -518,21 +537,106 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        let string_length = u16_vec.len() as u64;
+        if size <= string_length {
+            return Ok((false, string_length));
+        }
+
+        let this = self.eval_context_mut();
+
+        // Store the UTF-16 string.
+        let char_size = Size::from_bytes(2);
+        for (idx, c) in u16_vec.into_iter().chain(iter::once(0x0000)).enumerate() {
+            let place = this.mplace_field(mplace, idx as u64)?; 
+            this.write_scalar(Scalar::from_uint(c, char_size), place.into())?;
+        }
+        Ok((true, string_length))
+    }
+
+    /// Dispatches to appropriate implementations for allocating & writing OsString in Memory,
+    /// depending on the interpretation target.
+    fn alloc_os_str_as_target_str(
+        &mut self,
+        os_str: &OsStr,
+        memkind: MemoryKind<MiriMemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
+        let target_os = self.eval_context_ref().tcx.sess.target.target.target_os.as_str();
+        match target_os {
+            "linux" | "macos" => self.alloc_os_str_as_c_str(os_str, memkind),
+            "windows" => self.alloc_os_str_as_wide_str(os_str, memkind),
+            _ => throw_unsup_format!("OsString support for target OS not yet available"),
+        }
+    }
+
     fn alloc_os_str_as_c_str(
         &mut self,
         os_str: &OsStr,
         memkind: MemoryKind<MiriMemoryKind>,
-    ) -> Pointer<Tag> {
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
         let size = u64::try_from(os_str.len()).unwrap().checked_add(1).unwrap(); // Make space for `0` terminator.
         let this = self.eval_context_mut();
 
         let arg_type = this.tcx.mk_array(this.tcx.types.u8, size);
         let arg_place = this.allocate(this.layout_of(arg_type).unwrap(), memkind);
         self.write_os_str_to_c_str(os_str, arg_place.ptr, size).unwrap();
-        arg_place.ptr.assert_ptr()
+        Ok(arg_place.ptr.assert_ptr())
+    }
+
+    fn alloc_os_str_as_wide_str(
+        &mut self,
+        os_str: &OsStr,
+        memkind: MemoryKind<MiriMemoryKind>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
+        let size = os_str.len() as u64 + 1; // Make space for `0x0000` terminator.
+        let this = self.eval_context_mut();
+
+        let arg_type = this.tcx.mk_array(this.tcx.types.u16, size);
+        let arg_place = this.allocate(this.layout_of(arg_type).unwrap(), memkind);
+        self.write_os_str_to_wide_str(os_str, arg_place, size).unwrap();
+        Ok(arg_place.ptr.assert_ptr())
     }
 }
 
+#[cfg(target_os = "windows")]
+pub fn u16vec_to_osstring<'tcx>(u16_vec: Vec<u16>) -> InterpResult<'tcx, OsString> {
+    Ok(std::os::windows::ffi::OsStringExt::from_wide(&u16_vec[..]))
+}
+#[cfg(not(target_os = "windows"))]
+pub fn u16vec_to_osstring<'tcx>(u16_vec: Vec<u16>) -> InterpResult<'tcx, OsString> {
+    let s = String::from_utf16(&u16_vec[..])
+        .map_err(|_| err_unsup_format!("{:?} is not a valid utf-16 string", u16_vec))?;
+    Ok(s.into())
```

Comment:
I just realized we could probably use `Cow` here to avoid unnecessary allocations. :D Not hugely important, but a neat bit of Rust to play around with.

Reply:
To be honest, I'd need some time to learn using `Cow`.. 😅 
Can I try refactoring it to use `Cow` after PR #1256 lands to master?